Master vs. empty

pmitros opened the pull request:
@solashirai This is a PR which compares empty to current master, and will track master.

Commit:
first commit

Commit:
crowdxblock with basic functionality. can rate hints, make new hints, has default hints.

Commit:
test commit

Commit:
second test

Commit:
get_hint for non-default hints now works. hints for dummy "incorrect answer" are set so default hints dont appear.

Commit:
non-default hints now are correctly rated (better rating funciton to be added)

Commit:
not much changed from last commit

Commit:
changes made to how answer/hints show up (rating changes to be made)

Commit:
hinter now provides hints based on the mistake made for each dummy answer submission. After correct answer event takes place the incorrect answers and the hints provided for each one appear (up to 4 currently). Clicking on their orange box area will add 1 to the hint's rating for the respective answer, or you can submit your own hint (which will default to being under the first incorrect answer). Effectively you can only vote once, although it's not programmed to necessarily be so. sorry for long commit comment.

Commit:
added "Voted" to prevent multiple votes

Commit:
modifications made to correct hint giving/rating systems. now can clear all user side temporary dict/lists so that after you submit/vote on new hints you can go back and test changes. overall better.

Commit:
fixed to upvote when duplicate hint is provided

Commit:
minor changes

Commit:
version with error in trying to get ID. just uploading for reference.

Commit:
now have submission buttons that are specific to the incorrect answer to which the student wants to submit a hint. i believe all is functional.

Commit:
my computer will now not blow up after trying to get more hints than are available.

Commit:
mutliple hints show for voting after correct is reached.

Commit:
latest changes

Commit:
fixed mistakes I made with github doing stuff.

Commit:
most recent changes.

Commit:
Now properly deals with periods. Colon/Semicolon probably will mess things up still.

Commit:
no longer crash when hints run dry (gives "no more hints" instead)

Commit:
submit hint input appears correctly, default hints no longer show up for answers that don't have hints, all hints for answer show up (limit later)

Commit:
slight changes to hint showing during get_feedback

Commit:
changes to make colon/semicolon/quotes in hint submission not mess up the feedback (replace : with a word, for example). replacement words are not elegant but they aren't anything someon would type in so they will be fine for now i think.

Commit:
screenshot of hint

Commit:
screenshot of feedback

Commit:
many changes

Commit:
many changes

Commit:
changed to make the pictures easier to understand

Commit:
changes for first sprint

Commit:
added more comments to the code

Commit:
Update README.md

Commit:
Updated a few comments

Commit:
Delete README.md~

Commit:
Delete crowdxblock.py~

Commit:
Delete crowdxblockstudio.js~

Commit:
Delete crowdxblock.js~

Commit:
Delete crowdxblock.html~

Commit:
Delete crowdxblocka.html~

Commit:
Delete crowdxblockstudio.html~

Commit:
Delete .#crowdxblock.html

Commit:
Delete .#crowdxblock.js

Commit:
Delete crowdxblock.css~

Commit:
Create .gitignore

Commit:
Create tests.py

Commit:
Create test_crowdxblock.py

Commit:
added new files

Commit:
added defaults to hints to try to use tests and evaluate values

Commit:
made a simple test case, maybe correctly

Commit:
removed incorrect assertEqual

Commit:
updated some functions to remove self.Used

Commit:
Delete tests.py.orig

Commit:
removed clear_temp

Commit:
removed clear_temp

Commit:
removed HintsToUse

Commit:
fixed variable name

Commit:
fixed variable naming

Commit:
fixed variable names

Commit:
fixed hint rating

Commit:
fixing get_hint

Commit:
new course export

Commit:
fixed flagging hint

Commit:
fixed student hint submission, squashed bug

Commit:
show only the best hint

Commit:
fixed hint returns

Commit:
improving UX

Commit:
fixed mistakes

Commit:
variable renamings for javascript

Commit:
adding isstaff feedback

Commit:
fixing variables and such for isstaff feedback

Commit:
fixed variabl naming and staff feedback

Commit:
adding staff un-flag capabilities

Commit:
fixing js to make flagged hint section for staff

Commit:
removed print messages

Commit:
fixed JS method of showing hints

Commit:
fixed up up feedback, only for upvote at the moment

Commit:
flagged hint moderation functional

Commit:
fixed "list" to "len"

Commit:
fixed indentation

Commit:
added mustache

Commit:
fixed feedback bugs

Commit:
fixed unit switching with hinter

Commit:
fixed variables, working on css

Commit:
event logging, changes to feedback

Commit:
newly created hints show up in feedback

Commit:
fixed git issues

Commit:
saving

Commit:
renamed all "crowdxblock" to "crowdsourcehinter"

Commit:
implementing mustache templates

Commit:
moving to templates

Commit:
shifts to templates, incomplete

Commit:
faulty onStudentSubmission for hangout

Commit:
variable fixes and beginnings of Mustache templating

Commit:
attempt at mustache minimal example

Commit:
variable name fix

Commit:
trying to implement minimal example

Commit:
templates working

Commit:
templates implemented

Commit:
fixing hint flagging

Commit:
starting to figure out ux

Commit:
removed columns

Commit:
commit to keep track of hinter before moving into UX changes

Commit:
milestone

Commit:
commiting prior to more UX changes

Commit:
temporary push to github, possibly non functional version

Commit:
changes to flow of hinting/feedback mostly in place

Commit:
minimal parse_xml attempt

Commit:
added parsexml

Commit:
fixing up UX

Commit:
removed print/consolelogs

Commit:
further work on XML and fix to hint submission

Commit:
uptodate course export

Commit:
working on css changes

Commit:
fixed incorporation of initial hints

Commit:
added comment for temporarydict

Commit:
continuing css

Commit:
commit for meeting

Commit:
added examples

Commit:
removed double block in parsexml

Commit:
removed dict->string->dict patterns

Commit:
continuing various works

Commit:
mostly cleanup

Commit:
some css and html changes

Commit:
changed problem_graded to problem_check

Commit:
working on fixing problem_check and parse_xml

Commit:
parse_xml progress, fixing problem_graded/problem_check

Commit:
issues with logger.listen

Commit:
fixes to listening for events, errors exist with dict key/values being unicode

Commit:
touching up rough looks and unnecessary code

Commit:
changed "flagged" to "reported"

Commit:
typo fix

Commit:
small improvements

Commit:
removed unnecessary var

Commit:
code improvement

Commit:
renewed documentation

Commit:
naming changes (partial), fixes to Mustache template usage

Commit:
more naming fixes, removal of redundancies

Commit:
removed crowdsourcehinter egg-info

Commit:
removed vote var

Commit:
added .egg-info to gitignore

Commit:
removed unused funcitonality, fixed naming

Commit:
fixes/improvement to UX

Commit:
studio view/hint creation added

Commit:
added studio files

Commit:
removed prints

Commit:
improved studio hint editing (descriptions in UX, check that new hints are valid)

Commit:
various improvements, in progress

Commit:
test push

Commit:
updated old info

Commit:
sourced to source in naming

Commit:
testing

Commit:
updated readme

Commit:
added automatic setting for problem element

Commit:
general improvements

Commit:
updates

Commit:
corrected wording

Commit:
fixed staff rating, slight UX fixes

Commit:
slightly improved UX

Commit:
added new demo course

Commit:
fixed xml, fixed initial hint setting

Commit:
removed unnecessary code, fixed initial hint setting

Commit:
fixed UX for staff during hint contribution

Commit:
removed unnecessary class

Commit:
removed unnecessary code, changed explanations

Commit:
removed extra bracket from workbench scenario

Commit:
small code cleanup

pmitros committed:
Initial code review with comments about next steps

pmitros committed:
Cleans up decoding of student submissions

pmitros committed:
Moving Studio view to use JSON instead of AST, and to have slightly cleaner validation logic

pmitros committed:
Demo course replaced with something human-friendly and without extra stuff

Commit:
fixed minor issues. answers are not being stored correctly

Commit:
removed unused code

pmitros commented on crowdsourcehinter/crowdsourcehinter.py:
```diff
@@ -0,0 +1,409 @@
+    # (problem_element? hinting_element?). Trying to just change the
+    # name didn't seem to operate properly, check carefully what is
+    # changed
+    Element = String(default="", scope=Scope.content)
```
Incorrect case as well

pmitros replied:
Changed to `target_problem`

pmitros commented on crowdsourcehinter/crowdsourcehinter.py:
```diff
@@ -0,0 +1,409 @@
+        self.initial_hints = initial_hints
+        self.generic_hints = generic_hints
+        if len(str(data['element'])) > 1:
+            self.Element = str(data['element'])
```
Remove str (globally).

pmitros replied:
Removed all casts like this. Switched to bools where relevant.

pmitros commented on crowdsourcehinter/crowdsourcehinter.py:
```diff
@@ -0,0 +1,409 @@
+        frag.add_javascript_url('//cdnjs.cloudflare.com/ajax/libs/mustache.js/0.8.1/mustache.min.js')
+        frag.add_css(self.resource_string("static/css/crowdsourcehinter.css"))
+        frag.add_javascript(self.resource_string("static/js/src/crowdsourcehinter.js"))
+        frag.initialize_js('CrowdsourceHinter', {'hinting_element': self.Element, 'isStaff': get_user_is_staff})
```
Is this correct? (re: get_user_is_staff)

pmitros replied:
Yes it is. I don't like the API, but that's what it is

pmitros commented on crowdsourcehinter/crowdsourcehinter.py:
```diff
@@ -0,0 +1,409 @@
+        # there are no hints in hint_database.  this probably will
+        # occur only on the very first run of a unit containing this
+        # block.
+         if not self.hint_database:
```
Indentation is wrong. Does this work?

pmitros replied:
I'm also not sure if the code does the same thing...

pmitros replied:
I'd think through the logic here and break it into an initialization function

pmitros commented:
We need a mechanism to limit the size of this dict (and of each of the dict fields). (moving Ned's remaining comments over from https://github.com/solashirai/crowdsourcehinter/pull/2)

pmitros commented on crowdsourcehinter/crowdsourcehinter.py:
```diff
@@ -0,0 +1,409 @@
+        if (len(self.hint_database[answer]) - len(isreported)) > 0:
+            return 1
+        else:
+            return 0
```
Should return true or false (or have the int be meaningful -- e.g. # remaining hints)

pmitros commented on crowdsourcehinter/crowdsourcehinter.py:
```diff
@@ -0,0 +1,409 @@
+        for hint_key in self.hint_database[answer]:
+            for reported_keys in self.reported_hints:
+                if hint_key in self.reported_hints:
+                    isreported.append(hint_key)
```
Replace the inner loop with:

```
if hint_key in self.reported_hints: 
    isreported.append(hint_key)
```

pmitros replied:
Done.

pmitros commented on crowdsourcehinter/crowdsourcehinter.py:
```diff
@@ -0,0 +1,409 @@
+        answer_data = data['student_answer']
+        data_rating = data['student_rating']
+        data_hint = data['hint']
+        if data_hint == 'Sorry, there are no hints for this answer.':
```
This is fragile way to communicate between browser and server, and will definitely not work if we ever localize this code. It's better to use internal values to transmit state to the server, not human-readable text.

pmitros replied:
I left this in, but added a data field as well. Right now, neither is used anymore, I believe, but they will be used again in the future.

pmitros commented:
Most of the comments here are credit to @nedbat. This is a cleanup of https://github.com/solashirai/crowdsourcehinter/pull/2.

solashirai commented:
Thank you for the help.
At the moment I'm also looking into a problem with how student answers are received by the hinter. All student answers that contain spaces now show up with a + sign in place of spaces. I'm not sure if this is due to changes in the hinter or in the problem_check event. I'll replace this comment in the appropriate place after I look into it a bit more.
edit: i'll probably hold off changes until the weekend, this week is final exams for me.

pmitros commented:
That sounds like URL encoding. I think that's because of a change I made -- specifically a cleanup of extract_student_answers. The way this was handled before was a little bit funny and fragile. I didn't test that use case. Apologies.

nedbat commented:
Let me know when this PR is ready to review.  We need to find a revision to install with edx-platform.  if the code is churning too much, it will be difficult to settle on one.

pmitros commented:
@nedbat I stopped making changes in master at the time you asked to review. For some reason, the PR reviewed did not point into the head of master. My expectation, from the PR comments, is that we will want to:
- Potentially limit the growth of some of the data structures
- Fix the issue with spaces Sola identified
- Make the few minor/nit comments I carried over from the original PR

I will continue to work in branches on larger changes otherwise.

pmitros commented:
@nedbat Actually, if you give me a few hours, I'll make one more merge to master; I'm looking at my branch, and it does clean a few things up which I think will make the PR process a little bit nicer. I'd just like to validate it works. I'll let you know once that's done.

pmitros committed:
Code quality cleanup, following Ned's code review.

This was done in a branch pre-rebase. The branch didn't rebase cleanly. This was from a manual merge. More manual
(and automated) testing would be helpful.

Commit:
Merge pull request #5 from solashirai/pmitros/code-review-2

Code quality cleanup, following Ned's code review.

pmitros commented:
@nedbat I think I addressed all issues except for growing data structures. We're not launching this right now (we just want test cases), and if we were, they'd be unlikely to grow to anything substantial except by explicit attack (which we have a lot more issues with). I can put in caps for now if desired, but I'd rather hold off. At some point, we'd like to add code to guide students as to which hints to contribute (e.g. to places where there are few or no good hints already), and it would make sense to do it then as part of that subsystem.

solashirai commented:
@pmitros quite a few problems are actually popping up with the current changes. I will be able to start carefully looking at it tomorrow afternoon.

pmitros commented:
@solashirai It would be helpful to come up with a test strategy. The basic functionality works fine on my end, but it is hard to test for even moderate coverage without such.

pmitros commented:
@solashirai (It's something I'm thinking about, but test strategy is not an easy problem where XBlocks are)

Commit:
working on fixing answer extraction and initial settings

Commit:
fixed naming

Commit:
progress of fixing issues

Commit:
removed print

Commit:
Merge pull request #6 from solashirai/sola/working

Sola/working

Commit:
fixed issue with answers containing spaces

Commit:
removed unnecessary unquote

nedbat commented:
I've looked over my earlier comments as they stand in the code at 87b53fbe25f0a8e05ec0e42233cae509e5e9feef.

I don't see any attempt to limit the hints_database field, which was my primary concern.

I also notice discussion about url-quoting of answers.  Why aren't they being transmitted as JSON?

nedbat commented:
Also, you'll need to update the edx-platform PR with the correct hash.

Commit:
fixed naming, added documentation

Commit:
fixed typo

solashirai committed:
Merge pull request #7 from solashirai/sola/working

fixed issue with answers containing spaces

pmitros commented:
@nedbat Regarding limiting size, see my comment above. I'm happy to limit if really necessary, but it'd be a short-term hack.

pmitros committed:
Workbench compatibility tweaks

nedbat commented:
@pmitros I don't understand why a size limit would be a hack now.  Whatever guidance you give to students about what to hint, you will still need a limit to the data storage.  I understand the idea of getting the code out there to try it out, learn from it, etc.  I also understand the way software can live for a long time on production servers, get used in ways it was not intended, be found by people who don't understand its limitations, etc.  
Please implement a cap of some kind.  It should not be difficult: if the block is only used in ways that don't reach the cap, then it doesn't matter what items you discard when the cap is reached.

pmitros commented:
@nedbat It's not guidance. It's whether and what answers we ask them to give a hint for. 

Right now, we'd just like to get test cases in, actually, so the platform doesn't keep revolving underneath us. We're not launching this in any courses right now. 

Again, happy to do something now where it just clips or something, but overall, we'd be better building a heuristic for what to ask hints for. If an answer has 12 hints already, we don't need a 13th.

Commit:
removed unused code, fixed best_hint logic, attempt to limiting hint_database

nedbat commented:
Great, let's put in a limit that will clip.  Thanks.

Commit:
basic fixes

Commit:
added note of temporary solution

solashirai committed:
Merge pull request #8 from solashirai/sola/working

hint storage limits, fixed rating logic

pmitros commented:
@nedbat Sola put the limit in there. The code isn't too elegant, but it's a stopgap until we've built proper logic here. I'd really like to get the test cases into edx-platform.

nedbat commented:
@pmitros @solashirai Thanks, I appreciate the work you've put into this. :+1: Go ahead and update the edx-platform pull request with this hash: 518605f0a95190949fe77bd39158450639e2e1dc

solashirai commented:
@nedbat Already done. Thank you!

cclauss committed:
Create .travis.yml

cclauss committed:
[INCR-93] Run python-modernize on crowdsourcehinter

solashirai committed:
Merge pull request #10 from cclauss/INCR-93-python-modernize

[INCR-93] Run python-modernize on crowdsourcehinter

solashirai committed:
Merge pull request #9 from cclauss/patch-1

Create .travis.yml
